server/engine/src/main/java/org/eclipse/lsp/cobol/core/engine/TextTransformations.java

```diff
@@ -0,0 +1,213 @@
+        String replace;
+        Position start = eda.getPosition();
+        if (extensions.containsKey(currentRange)) {
+          TextTransformations textTransformations = extensions.get(currentRange);
+          replace = textTransformations.calculateExtendedText();
+          extToOriginalMap.putAll(shift(eda.getPosition(), textTransformations.getExtToOriginalMap()));
+        } else {
+          replace = replacements.get(currentRange);
+        }
+        eda.append(replace);
+        Position end = eda.getPosition();
+        Range extRange = new Range(start, end);
+        if (!isRangeEmpty(extRange)) {
+          extToOriginalMap.put(extRange, new Location(uri, currentRange));
+        }
+        lineNumber = currentRange.getEnd().getLine();
+        prevRange = currentRange;
+        currentRange = ranges.isEmpty() ? null : ranges.removeFirst();
+      }
+    }
+    for (int i = lineNumber; i < lines.length; i++) {
+      if (prevRange != null && prevRange.getEnd().getLine() == i) {
+        eda.append(lines[i], prevRange.getEnd().getCharacter(), lines[i].length());
+      } else {
+        eda.append(lines[i]);
+      }
+    }
+    return eda.toString();
+  }
+
+  private Map<Range, Location> shift(Position extPos, Map<Range, Location> extToOriginalMap) {
+    Map<Range, Location> result = new HashMap<>();
+    extToOriginalMap.forEach(
+        (k, v) -> {
+          Range newKey =
+              new Range(
+                  new Position(
+                      k.getStart().getLine() + extPos.getLine(), k.getStart().getCharacter()),
+                  new Position(k.getEnd().getLine() + extPos.getLine(), k.getEnd().getCharacter()));
+          result.put(newKey, v);
+        });
+    return result;
+  }
+
+  private boolean isRangeEmpty(Range range) {
+    return range.getStart().equals(range.getEnd());
+  }
+
+  /**
+   * Get a map of extended document. It will be populated after calculateExtendedText call
+   *
+   * @return a map of extended document.
+   */
+  public Map<Range, Location> getExtendedDocumentMap() {
+    return extToOriginalMap;
+  }
+
+  /**
+   * Replace copy statement with result of copybook substitution
+   *
+   * @param copyNode node representation of copybook
+   * @param textTransformations Copybook's transformations
+   */
+  public void extend(CopyNode copyNode, TextTransformations textTransformations) {
+    copyNodes.add(copyNode);
+    Range range = copyNode.getLocality().getRange();
+    range =
+        new Range(
+            new Position(range.getStart().getLine(), 0),
+            new Position(range.getEnd().getLine(), range.getEnd().getCharacter()));
+    extensions.put(range, textTransformations);
+  }
+
+  /**
+   * Substitute location of original document with a new content.
+   *
+   * @param range location
+   * @param newText new content
+   */
+  public void replace(Range range, String newText) {
```

Comment:
that method is too often being called with newText = "",
so do consider to add a new method to remove all the text from that range.